## SNPs analysis pipeline.
Performs unsupervised (PCA and CA) and supervised (Random Forests) on the potentially deleterious SNPs data. Parses intra_SNPs and btw_SNPs. Filter SNPs that produce premature stop codons by one side and non-synonymous SNPs by other. Runs PROVEAN to determine the potentially deleterious non-synonymous SNPs and builds a deleterious SNPs dataset by merging the stop gain codon SNPs and the PROVEAN significative ones, indicating the proportion of the isolates within each lineage harboring each SNP. A random sample of isolates is selected for each lineage and VCF files generated by the pipeline in mapsPipe are used to determine if the isolate has the potentially deleterious SNP or not. These random samples are analyzed by unsupervised and supervised methods to determine the variables more important in the classification. Then the genes that have potentially deleterious SNPs present in more than the 85% of the isolates of each lineage are written in TXT files. Separated files for stop gain SNPs and to PROVEAN significative SNPs.<br>
Run launch_snpsPipe.sh for use, input are intra and btw SNPs files generated, and VCF files generated in mapsPipe. Needs the installation of PROVEAN (http://provean.jcvi.org/index.php)